File: sdk/csharp/src/Providers/InboxKitten.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text.Json.Nodes;

namespace XxxXTeam.TempMail.Providers;

/// <summary>InboxKitten 渠道（inboxkitten.com/api/v1/mail）。</summary>
public static class InboxKitten
{
    private const string ApiBase = "https://inboxkitten.com/api/v1/mail";
    private const string Domain = "inboxkitten.com";
    private static readonly Random Rand = new();
    private static readonly Dictionary<string, string> HeadersJson = new() { ["Accept"] = "application/json" };
    private static readonly Dictionary<string, string> HeadersHtml = new() { ["Accept"] = "text/html,*/*" };

    private static string RandomLocal()
    {
        const string chars = "abcdefghijklmnopqrstuvwxyz0123456789";
        var sb = new System.Text.StringBuilder("sdk");
        for (var i = 0; i < 16; i++) sb.Append(chars[Rand.Next(chars.Length)]);
        return sb.ToString();
    }

    public static EmailInfo Generate() => new("inboxkitten", $"{RandomLocal()}@{Domain}");

    public static List<Email> GetEmails(string email)
    {
        email = (email ?? "").Trim();
        var local = email.Split('@')[0];
        if (string.IsNullOrEmpty(local)) throw new Exception("inboxkitten: empty email");

        var resp = Http.Get($"{ApiBase}/list?recipient={WebUtility.UrlEncode(local)}", HeadersJson);
        resp.EnsureSuccess();
        var rows = Json.Parse(resp.Body) as JsonArray;
        var result = new List<Email>();
        if (rows is null) return result;
        foreach (var item in rows)
            if (item is JsonObject row)
                result.Add(Normalize.NormalizeEmail(DetailRaw(row, email), email));
        return result;
    }

    private static Dictionary<string, object?> DetailRaw(JsonObject row, string recipient)
    {
        var storage = row["storage"] as JsonObject;
        var region = (storage?["region"]?.GetValue<string>() ?? "").Trim();
        var key = (storage?["key"]?.GetValue<string>() ?? "").Trim();
        var message = row["message"] as JsonObject;
        var headers = message?["headers"] as JsonObject;
        var envelope = row["envelope"] as JsonObject;

        var raw = new Dictionary<string, object?>
        {
            ["id"] = key,
            ["messageId"] = key,
            ["from"] = FirstNonEmpty(headers?["from"]?.GetValue<string>(), envelope?["sender"]?.GetValue<string>()),
            ["to"] = FirstNonEmpty(row["recipient"]?.GetValue<string>(), recipient),
            ["subject"] = headers?["subject"]?.GetValue<string>() ?? "",
        };
        if (string.IsNullOrEmpty(region) || string.IsNullOrEmpty(key)) return raw;

        try
        {
            var metaResp = Http.Get($"{ApiBase}/getKey?region={WebUtility.UrlEncode(region)}&key={WebUtility.UrlEncode(key)}", HeadersJson);
            metaResp.EnsureSuccess();
            var meta = Json.Parse(metaResp.Body) as JsonObject;
            var htmlResp = Http.Get($"{ApiBase}/getHtml?region={WebUtility.UrlEncode(region)}&key={WebUtility.UrlEncode(key)}", HeadersHtml);
            htmlResp.EnsureSuccess();
            var html = htmlResp.Body;
            raw["from"] = FirstNonEmpty(meta?["name"]?.GetValue<string>(), raw["from"] as string);
            raw["subject"] = FirstNonEmpty(meta?["subject"]?.GetValue<string>(), raw["subject"] as string);
            raw["text"] = Normalize.HtmlToText(html);
            raw["html"] = html;
        }
        catch { /* 保留摘要 */ }
        return raw;
    }

    private static string FirstNonEmpty(params string?[] values)
    {
        foreach (var v in values) if (!string.IsNullOrWhiteSpace(v)) return v!;
        return "";
    }
}

/// <summary>1SecMail 渠道（tmaily.com，会话 Cookie 鉴权）。</summary>
public static class OneSecMail
{
    private const string BaseUrl = "https://tmaily.com/";
    private static readonly Dictionary<string, string> Headers = new()
    {
        ["Accept"] = "application/json",
        ["User-Agent"] = "Mozilla/5.0",
    };

    public static EmailInfo Generate()
    {
        var resp = Http.Get($"{BaseUrl}generate", Headers);
        resp.EnsureSuccess();
        var cookie = ExtractCookie(resp);
        if (string.IsNullOrEmpty(cookie)) throw new Exception("1sec-mail: 未获取到会话 Cookie");
        var email = Json.Str(Json.Parse(resp.Body), "address").Trim();
        if (!email.Contains('@')) throw new Exception("1sec-mail: 无效的邮箱响应");
        return new EmailInfo("1sec-mail", email, cookie);
    }

    private static string ExtractCookie(HttpResult resp)
    {
        foreach (var c in resp.SetCookies)
        {
            var idx = c.IndexOf("TMaily_sid=", StringComparison.Ordinal);
            if (idx < 0) continue;
            var seg = c[idx..];
            var semi = seg.IndexOf(';');
            return semi < 0 ? seg : seg[..semi];
        }
        return "";
    }

    public static List<Email> GetEmails(string? token, string email)
    {
        var address = (email ?? "").Trim();
        if (string.IsNullOrEmpty(address)) throw new Exception("1sec-mail: 邮箱地址为空");
        var headers = new Dictionary<string, string>(Headers) { ["Cookie"] = token ?? "" };
        var resp = Http.Get($"{BaseUrl}emails?address={WebUtility.UrlEncode(address)}", headers);
        resp.EnsureSuccess();
        var rows = Json.Parse(resp.Body) as JsonArray;
        var result = new List<Email>();
        if (rows is null) return result;
        foreach (var item in rows)
            if (item is JsonObject) result.Add(Normalize.NormalizeEmail(Json.ToDict(item), address));
        return result;
    }
}

File: sdk/php/src/Providers/InboxKitten.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;

/**
 * InboxKitten 渠道实现 — https://inboxkitten.com
 *
 * 本地生成随机邮箱，列表 GET /api/v1/mail/list?recipient={local}，
 * 逐封通过 storage 的 region/key 调用 getKey（元数据）+ getHtml（正文）补全。
 */
final class InboxKitten
{
    private const CHANNEL = 'inboxkitten';
    private const API_BASE = 'https://inboxkitten.com/api/v1/mail';
    private const DOMAIN = 'inboxkitten.com';

    /** @var array<string,string> */
    private const HEADERS_JSON = ['Accept' => 'application/json'];

    /** @var array<string,string> */
    private const HEADERS_HTML = ['Accept' => 'text/html,*/*'];

    private static function randomLocal(): string
    {
        $chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
        $out = 'sdk';
        for ($i = 0; $i < 16; $i++) {
            $out .= $chars[random_int(0, strlen($chars) - 1)];
        }
        return $out;
    }

    public static function generate(): EmailInfo
    {
        return new EmailInfo(self::CHANNEL, self::randomLocal() . '@' . self::DOMAIN);
    }

    /**
     * 获取单封元数据。
     *
     * @return array<mixed>
     */
    private static function fetchMeta(string $region, string $key): array
    {
        $resp = HttpClient::get(
            self::API_BASE . '/getKey?region=' . rawurlencode($region) . '&key=' . rawurlencode($key),
            self::HEADERS_JSON,
        );
        if ($resp->getStatusCode() >= 400) {
            throw new \RuntimeException('inboxkitten: getKey 失败 ' . $resp->getStatusCode());
        }
        return HttpClient::json($resp);
    }

    private static function fetchHtml(string $region, string $key): string
    {
        $resp = HttpClient::get(
            self::API_BASE . '/getHtml?region=' . rawurlencode($region) . '&key=' . rawurlencode($key),
            self::HEADERS_HTML,
        );
        if ($resp->getStatusCode() >= 400) {
            throw new \RuntimeException('inboxkitten: getHtml 失败 ' . $resp->getStatusCode());
        }
        return (string) $resp->getBody();
    }

    /**
     * 将列表条目转换为标准化前的原始字段，尽力补全正文。
     *
     * @param array<mixed> $row
     * @return array<mixed>
     */
    private static function detailRaw(array $row, string $recipient): array
    {
        $storage = is_array($row['storage'] ?? null) ? $row['storage'] : [];
        $message = is_array($row['message'] ?? null) ? $row['message'] : [];
        $headers = is_array($message['headers'] ?? null) ? $message['headers'] : [];
        $envelope = is_array($row['envelope'] ?? null) ? $row['envelope'] : [];
        $region = trim((string) ($storage['region'] ?? ''));
        $key = trim((string) ($storage['key'] ?? ''));

        $raw = [
            'id' => $key,
            'messageId' => $key,
            'from' => $headers['from'] ?? ($envelope['sender'] ?? ''),
            'to' => $row['recipient'] ?? ($envelope['targets'] ?? $recipient),
            'subject' => $headers['subject'] ?? '',
            'timestamp' => $row['timestamp'] ?? null,
        ];
        if ($region === '' || $key === '') {
            return $raw;
        }

        try {
            $meta = self::fetchMeta($region, $key);
            $html = self::fetchHtml($region, $key);
            $raw['from'] = $meta['name'] ?? $raw['from'];
            $raw['to'] = $meta['recipients'] ?? $raw['to'];
            $raw['subject'] = $meta['subject'] ?? $raw['subject'];
            $raw['html'] = $html;
        } catch (\Throwable) {
            // 补全失败时退回列表元数据
        }
        return $raw;
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email): array
    {
        $local = explode('@', trim($email))[0];
        if ($local === '') {
            throw new \InvalidArgumentException('inboxkitten: empty email');
        }
        $resp = HttpClient::get(
            self::API_BASE . '/list?recipient=' . rawurlencode($local),
            self::HEADERS_JSON,
        );
        if ($resp->getStatusCode() >= 400) {
            throw new \RuntimeException('inboxkitten: 列表请求失败 ' . $resp->getStatusCode());
        }
        $rows = HttpClient::json($resp);
        if (!array_is_list($rows)) {
            return [];
        }
        $out = [];
        foreach ($rows as $item) {
            if (is_array($item)) {
                $out[] = Normalize::email(self::detailRaw($item, $email), $email);
            }
        }
        return $out;
    }
}
